presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -0,0 +1,525 @@
+import static org.apache.iceberg.TableMetadata.newTableMetadata;
+import static org.apache.iceberg.Transactions.createTableTransaction;
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+public class IcebergMetadata
+        implements ConnectorMetadata
+{
+    private final HiveMetastore metastore;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final JsonCodec<CommitTaskData> commitTaskCodec;
+
+    private Transaction transaction;
+
+    public IcebergMetadata(
+            HiveMetastore metastore,
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            JsonCodec<CommitTaskData> commitTaskCodec)
+    {
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.commitTaskCodec = requireNonNull(commitTaskCodec, "commitTaskCodec is null");
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        return metastore.getAllDatabases();
+    }
+
+    @Override
+    public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        IcebergTableHandle handle = IcebergTableHandle.from(tableName);
+        Optional<Table> table = metastore.getTable(handle.getSchemaName(), handle.getTableName());
+        if (!table.isPresent()) {
+            return null;
+        }
+        if (!isIcebergTable(table.get())) {
+            throw new PrestoException(NOT_SUPPORTED, "Not an Iceberg table: " + tableName);
+        }
+        return handle;
+    }
+
+    @Override
+    public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        IcebergTableHandle table = IcebergTableHandle.from(tableName);
+        if (table.getTableType() == TableType.PARTITIONS) {
+            Configuration configuration = getConfiguration(session, tableName.getSchemaName());
+            org.apache.iceberg.Table icebergTable = getIcebergTable(table.getSchemaName(), table.getTableName(), configuration, metastore);
+            return Optional.of(new PartitionTable(table, session, typeManager, icebergTable));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public ConnectorTableProperties getTableProperties(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        return new ConnectorTableProperties();
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        return getTableMetadata(session, ((IcebergTableHandle) table).getSchemaTableName());
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
```

Comment:
Agreed. Doing this will require some metastore API changes. To do this efficiently, we'll need to a method like `getAllTablesWithProperty(databaseName, propertyName, propertyValue)`, basically a generic version of `getAllViews` (see `ThriftHiveMetastore.getPrestoViews`). I'll add a TODO.